---
title: "Patterns of microbial community assembly"
output: html_document
---

```{r setup.1, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, cache = TRUE, message = FALSE)
```

```{r setup.2, warning = FALSE, include = FALSE, cache = TRUE}
library(data.table)
library(ggmap)
library(ggplot2)
library(ggplotify)
library(grid)
library(gridExtra)
library(TDA)
library(TDAstats)
library(phyloseq)
library(phylosmith)
library(plotly)
library(reshape2)
library(stringr)
library(tidyverse)
library(vegan)

source('./src/pommier.R')

theme_set(theme_light())

```

An overview of community structures in microbial ecology.

Borrowing heavily from [Partitioning the turnover and nestedness components of beta diversity](https://onlinelibrary.wiley.com/doi/full/10.1111/j.1466-8238.2009.00490.x) by Andres Baselga.

## Simulating communities

Start off with basic examples from Baselga's paper. Starting by reproducing these structures:

![alt](./images/baselga-structures.png)

## Situation A: Nestedness

```{r baselga_sim.A}
site_A1 <- rep(1, 12)
site_A2 <- sim.sample.contig_presence(12, 4)
site_A3 <- sim.sample.contig_presence(12, 2)

sim_A.otu_table <- otu_mat_from_lists(site_A1, site_A2, site_A3)

sim_A.tax_mat <- sim.sample.tax_info(nrow(sim_A.otu_table))

sim_A.sample_data <- sim.sample.treatment_data(sim_A.otu_table)

sim_A.physeq <- phyloseq(otu_table(sim_A.otu_table, taxa_are_rows = TRUE), 
                         tax_table(sim_A.tax_mat),
                         sim_A.sample_data)


```


```{r baselga_sim.A.plot}
plot_communities(sim_A.physeq@otu_table) + 
  labs(title = "Simulation A: Nestedness",
       subtitle = "The observed microbiome of Site A3 (resp. A2) is a proper subset of the observed microbiome \n in Site A2 (resp. A1)")
```


```{r baselga_sim.A.hom}
sim_A.homology <- get_homology_from_phyloseq(sim_A.physeq)
plot_barcode(sim_A.homology)
```


## Situation B: Spatial turnover



```{r}
site_B1 <- sim.sample.contig_presence(12, 6)
site_B2 <- c(sim.sample.contig_presence(6, 3), sim.sample.contig_presence(6, 3))
site_B3 <- c(sim.sample.contig_presence(9, 3), rep(1, 3))

sim_B.otu_table <- otu_mat_from_lists(site_B1, site_B2, site_B3)

sim_B.tax_mat <- sim.sample.tax_info(nrow(sim_B.otu_table))

sim_B.sample_data <- sim.sample.treatment_data(sim_B.otu_table)

sim_B.physeq <- phyloseq(otu_table(sim_B.otu_table, taxa_are_rows = TRUE), 
                         tax_table(sim_B.tax_mat),
                         sim_B.sample_data)

```

```{r}
plot_communities(sim_B.otu_table)
```



## Increasing complexity

Will first start off by increasing number of samples and OTUs.

First simulation - everything present in all samples:

```{r}

num_samples = 30
num_otus = 10

sim_1.otu_table <- matrix(sample(1:100, num_samples * num_otus, replace = TRUE), 
                          nrow = num_otus, ncol = num_samples,
                          dimnames = list(paste("OTU", 1:num_otus, sep = "_"), 
                                          paste("Sample", 1:num_samples, sep = "_")))
head(sim_1.otu_table[1:3, 1:10])
```

```{r}
plot_communities(sim_1.otu_table)
```

Adding fake taxonomy information:

```{r}
sim_1.tax_mat = matrix(sample(letters, 70, replace = TRUE), nrow = nrow(sim_1.otu_table), ncol = 7)
rownames(sim_1.tax_mat) <- rownames(sim_1.otu_table)
colnames(sim_1.tax_mat) <- c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species")
sim_1.tax_mat
```


Adding fake sample data
```{r}
sim_1.sample_data <- sample_data(data.frame(
  Treatment = rep(c("Group 1", "Group 2"), each = 200),
  row.names = colnames(sim_1.otu_table),
  stringsAsFactors = FALSE))
sim_1.sample_data
```



```{r}
sim_1.physeq <- phyloseq(otu_table(sim_1.otu_table, taxa_are_rows = TRUE), 
                         tax_table(sim_1.tax_mat),
                         sim_1.sample_data)
sim_1.physeq
```
